### PR TITLE
add nll_loss_forward & nll_loss_backward kernel and op

### PR DESCRIPTION
**Pull request by AlienLiang23**
add nll_loss_forward & nll_loss_backward kernel and op

### CODE REVIEW

**Comment by AlienLiang23** [comment]
image
image
four failed UTs are XFAIL in CUDA

**Comment by AlienLiang23** [comment]
image
CUDA has same error with enforcing this case not skipped

**Comment by AlienLiang23** [comment]
image
image
test_ops al passed except skipped ones

**Comment by fengyuan14** [comment]
Hold on merge. According to @chunhuanMeng's comments, the operators fail on cases of test_nn.py and test_modules.py from IPEX report. Will add test_nn.py and test_modules.py first and don't merge until all bugs are fixed.

File: src/aten/LossNLL.cpp [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+      self,
+      target,
+      ((weight.has_value() && (*weight).defined())
+           ? at::OptionalTensorRef(*weight)
+           : at::OptionalTensorRef()),
+      reduction,
+      ignore_index,
+      output,
+      total_weight);
+  return native::xpu::launch_nll_loss_forward_kernel(
```
Comment (fengyuan14) [inline src/aten/LossNLL.cpp:98]:
```suggestion
  return native::xpu::nll_loss_forward_kernel(
```

File: src/aten/sycl/LossNLLKernel.h [hunk 2]
```diff
@@ -0,0 +1,25 @@
+#pragma once
+#include <ATen/native/TensorIterator.h>
+
+namespace at::native::xpu {
+
+std::tuple<Tensor&, Tensor&> launch_nll_loss_forward_kernel(
```
Comment (fengyuan14) [inline src/aten/sycl/LossNLLKernel.h:6]:
```suggestion
std::tuple<Tensor&, Tensor&> nll_loss_forward_kernel(
```

File: src/aten/sycl/LossNLLKernel.h [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    const Tensor& self,
+    const Tensor& target,
+    const OptionalTensorRef weight_opt,
+    int64_t reduction,
+    int64_t ignore_index,
+    Tensor& output,
+    Tensor& total_weight);
+
+Tensor& launch_nll_loss_backward_kernel(
```
Comment (fengyuan14) [inline src/aten/sycl/LossNLLKernel.h:15]:
```suggestion
Tensor& nll_loss_backward_kernel(
```

File: src/aten/LossNLL.cpp [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+      self,
+      target,
+      ((weight.has_value() && (*weight).defined())
+           ? at::OptionalTensorRef(*weight)
+           : at::OptionalTensorRef()),
+      reduction,
+      ignore_index,
+      total_weight,
+      grad_input);
+  return native::xpu::launch_nll_loss_backward_kernel(
```
Comment (fengyuan14) [inline src/aten/LossNLL.cpp:220]:
```suggestion
  return native::xpu::nll_loss_backward_kernel(
```

File: src/aten/LossNLL.cpp [hunk 5]
```diff
@@ -0,0 +1,252 @@
+#include <ATen/ATen.h>
+#include <ATen/XPUNativeFunctions.h>
+#include <ATen/core/Reduction.h>
+#include <ATen/core/Tensor.h>
+#include <ATen/core/op_registration/adaption.h>
+#include <aten/sycl/LossNLLKernel.h>
+#include <comm/RegisterUtils.h>
+
+namespace at {
+void meta_func_nll_loss_forward(
```
Comment (fengyuan14) [inline src/aten/LossNLL.cpp:10]:
```suggestion
void nll_loss_forward_meta(
```

File: src/aten/LossNLL.cpp [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+      common_device, output, "xpu::nll_loss_forward_out", "output");
+  c10::impl::check_and_update_common_device(
+      common_device, total_weight, "xpu::nll_loss_forward_out", "total_weight");
+  c10::impl::check_and_update_common_device(
+      common_device, self, "xpu::nll_loss_forward_out", "self");
+  c10::impl::check_and_update_common_device(
+      common_device, target, "xpu::nll_loss_forward_out", "target");
+  c10::impl::check_and_update_common_device(
+      common_device, weight, "xpu::nll_loss_forward_out", "weight");
+  meta_func_nll_loss_forward(
```
Comment (fengyuan14) [inline src/aten/LossNLL.cpp:88]:
```suggestion
  nll_loss_forward_meta(
```

File: src/aten/LossNLL.cpp [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+    const c10::optional<Tensor>& weight,
+    int64_t reduction,
+    int64_t ignore_index) {
+  Tensor output;
+  Tensor total_weight;
+  return nll_loss_forward_out(
+      self, target, weight, reduction, ignore_index, output, total_weight);
+}
+
+void meta_func_nll_loss_backward(
```
Comment (fengyuan14) [inline src/aten/LossNLL.cpp:122]:
```suggestion
void nll_loss_backward_meta(
```

File: src/aten/LossNLL.cpp [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+  c10::impl::check_and_update_common_device(
+      common_device, target, "xpu::nll_loss_backward_out", "target");
+  c10::impl::check_and_update_common_device(
+      common_device, weight, "xpu::nll_loss_backward_out", "weight");
+  c10::impl::check_and_update_common_device(
+      common_device,
+      total_weight,
+      "xpu::nll_loss_backward_out",
+      "total_weight");
+  meta_func_nll_loss_backward(
```
Comment (fengyuan14) [inline src/aten/LossNLL.cpp:209]:
```suggestion
  nll_loss_backward_meta(
```

**Review by fengyuan14 — APPROVED** [review]

**Pull request merged**